src/daemon/daemon.cpp

```diff
@@ -2668,6 +2680,82 @@ catch (const std::exception& e)
+                [this, destination_name]() noexcept -> void { release_resources(destination_name); });
+
+            assert(instance_trail.index() == 0);
+            const auto source_vm_ptr = std::get<0>(instance_trail)->second;
+            assert(source_vm_ptr);
+            const VirtualMachine::State source_vm_state = source_vm_ptr->current_state();
+            if (source_vm_state != VirtualMachine::State::stopped && source_vm_state != VirtualMachine::State::off)
+            {
+                throw std::runtime_error("Please stop instance " + source_name + " before you clone it.");
+            }
+
+            // signal that the new instance is being cooked up
+            preparing_instances.insert(destination_name);
+            auto& src_spec = vm_instance_specs[source_name];
+            auto dest_spec = clone_spec(src_spec, source_name, destination_name);
+
+            config->vault->clone(source_name, destination_name);
+
+            const mp::VMImage dest_vm_image = fetch_image_for(destination_name, *config->factory, *config->vault);
+
+            // QemuVirtualMachine constructor depends on vm_instance_specs[destination_name], so the appending of the
+            // dest_spec has to be done before the function create_vm_and_clone_instance_dir_data
+            vm_instance_specs.emplace(destination_name, dest_spec);
+            operative_instances[destination_name] =
+                config->factory->create_vm_and_clone_instance_dir_data(src_spec,
+                                                                       dest_spec,
+                                                                       source_name,
+                                                                       destination_name,
+                                                                       dest_vm_image,
+                                                                       *config->ssh_key_provider,
+                                                                       *this);
+            ++src_spec.clone_count;
+            // preparing instance is done
+            preparing_instances.erase(destination_name);
+            persist_instances();
+            init_mounts(destination_name);
+
+            CloneReply rpc_response;
+            rpc_response.set_reply_message(fmt::format("Cloned from {} to {}.\n", source_name, destination_name));
+            server->Write(rpc_response);
+            rollback_clean_up_all_resource_of_dest_instance.dismiss();
+        }
+        status_promise->set_value(status);
+    }
+    catch (const mp::CloneInvalidNameException& e)
+    {
+        // all CloneInvalidNameException throws in generate_destination_instance_name_for_clone
+        status_promise->set_value(grpc::Status(grpc::StatusCode::INVALID_ARGUMENT, e.what()));
+    }
+    catch (const std::runtime_error& e)
+    {
+        status_promise->set_value(grpc::Status(grpc::StatusCode::FAILED_PRECONDITION, e.what()));
+    }
```

Comment:
```suggestion
    catch (const std::exception& e)
    {
        status_promise->set_value(grpc::Status(grpc::StatusCode::INTERNAL, e.what()));
    }
```

`std::runtime_error` is a child of `std::exception` meaning that other exception types could crash the daemon.

Reply:
So basically we have three types error `grpc::StatusCode` here, `grpc::StatusCode::FAILED_PRECONDITION`, `grpc::StatusCode::INVALID_ARGUMENTS` and `grpc::StatusCode::INTERAL`. That seems to be more fine-grained.